assets/prototype/application/ui/components/display/entityListFilterBar/useEntityListFilterState.js

```diff
@@ -0,0 +1,89 @@
+import { useReducer, useEffect } from '@wordpress/element';
+
+const useEntityListFilterState = () => {
+	const initialState = {
+		perPage: 6,
+		searchText: '',
+		showEntityFilters: false,
+		view: 'list',
+	};
+	const [state, dispatch] = useReducer(reducer, initialState);
+
+	useEffect(() => {
+		console.log('useEntityListFilterState >>>', state);
+	}, [state]);
+
+	const setGridView = () => {
+		dispatch({
+			type: 'SET_VIEW',
+			view: 'grid',
+		});
+	};
+
+	const setListView = () => {
+		dispatch({
+			type: 'SET_VIEW',
+			view: 'list',
+		});
+	};
+
+	const setPerPage = (perPage) => {
+		dispatch({
+			perPage,
+			type: 'SET_PER_PAGE',
+		});
+	};
+
+	const setSearchText = (searchText) => {
+		dispatch({
+			searchText,
+			type: 'SET_SEARCH_TEXT',
+		});
+	};
+
+	const toggleEntityFilters = () => {
+		dispatch({
+			type: 'TOGGLE_ENTITY_FILTERS',
+		});
+	};
+
+	const perPage = state.perPage;
+	const searchText = state.searchText;
+	const showEntityFilters = state.showEntityFilters;
+	const view = state.view;
+
+	return {
+		perPage,
+		searchText,
+		setGridView,
+		setListView,
+		setPerPage,
+		setSearchText,
+		showEntityFilters,
+		toggleEntityFilters,
+		view,
+	};
+};
+
+const reducer = (state, action) => {
+	const { perPage, searchText, view } = action;
+
+	switch (action.type) {
+		case 'SET_PER_PAGE':
+			return { ...state, perPage };
+
+		case 'SET_SEARCH_TEXT':
+			return { ...state, searchText };
+
+		case 'SET_VIEW':
+			return { ...state, view };
+
+		case 'TOGGLE_ENTITY_FILTERS':
+			return { ...state, showEntityFilters: !state.showEntityFilters };
```

Comment:
do you already have a plan in mind for where the actual list filtering will occur?

I see there being two options the way things are set up now:

##### Option A - External Filtering
- filtering is done higher up the tree and the properties being controlled by the filter bar, like `perPage`, are pulled into that point in the tree and then used to filter the list directly
- so this would be doing absolutely nothing more than changing the filter "state" and the actual filtering is the responsibility of another component/service
- this might create a disconnect between this reducer and the actual filter logic though

##### Option B - Internal Filtering
- a set of callbacks are supplied to this hook when it is first called, like:

```js
const useEntityListFilterState = ({
	perPageFilter,
	searchTextFilter,
	viewFilter,
	toggleShowFilters,
}) => {
```

and those callbacks are triggered when the corresponding action is called:

```js
	switch (action.type) {
		case 'SET_PER_PAGE':
			perPageFilter(perPage );
			return { ...state, perPage };

		case 'SET_SEARCH_TEXT':
			searchTextFilter(searchText );
			return { ...state, searchText };

		case 'SET_VIEW':
			viewFilter(view );
			return { ...state, view };

		case 'TOGGLE_ENTITY_FILTERS':
			const showFilters = !state.showEntityFilters;
			toggleShowFilters(showFilters );
			return { ...state, showEntityFilters: showFilters  };
```

- this might make it easier to see where/how things are actually working, but also might be putting too much responsibility on this logic

thoughts?

Reply:
I think Option A is cleaner.